# Population Statistics with Stacks

[Setup Stacks](#setup-stacks) \
[Input files](#input-files) \
[populations command](#populations-command) \
[populations output](#populations-output) \
[Run populations: between islands](#run-populations-between-islands) \
[Run populations: within Clemente sites](#run-populations-within-clemente-sites) \
[Run populations: within Catalina sites](#run-populations-within-catalina-sites)

## Setup Stacks 

Download here: https://catchenlab.life.illinois.edu/stacks/ \
Install instructions here: https://catchenlab.life.illinois.edu/stacks/manual/#install
    
Commands are run through terminal from the stacks program directory when on the lab Desktop:

    cd ~/Desktop/stacks-2.60

Or if you have a stacks conda environment installed:

    conda activate stacks-2.68
    
Test that stacks is working properly, if it is, this next line will give you the usage instructions for `populations` command in stacks

    populations --h

## Input files:

Before we start running things, we need two input files and the path of where they live. 
1) the vcf file (output from stacks) \
Go to the directory where your .vcf file is and `pwd` to get the correct path 
2) population codes file for the files included in the vcf \
The pop_codes.csv file should be 2 columns; the name of the files (taken from the final stats file in ipyrad outfiles) and the site designations

ex: `ClemSIFI123a_stacks_pop_codes.csv`

		Sibara_filifolia_AJC_9654cat_SCl_1	Zone_1
		Sibara_filifolia_AJC_9656cat_SCl_1	Zone_1
		Sibara_filifolia_AJC_Z1_2_9655_SCl_1	Zone_1
		Sibara_filifolia_KHL4798_Z2_9659_SCl_2	Zone_2
		Sibara_filifolia_KHL4802_Z2_10269_SCl_2	Zone_2
		Sibara_filifolia_KHLM36_9645cat_SCl_4	Zone_4
		Sibara_filifolia_KHLML12_10279cat_SCl_4	Zone_4
		Sibara_filifolia_KHLML13_10280cat_SCl_4	Zone_4
		Sibara_filifolia_KHL_297_ML15_Z5_10115_SCl_5	Zone_5
		Sibara_filifolia_KHL_297_ML17_Z5_10116_SCl_5	Zone_5

### populations command
You can `populations -h` to see usage / all the possible flags/parameters for the `populations` command and their explanations.
Plug in the correct paths for your vcf, output directory, and pop_codes.csv.

	populations -V PATH/TO/VCF.vcf -O PATH/TO/OUTPUT/DIRECTORY -M PATH/TO/POP_CODES.csv --fstats --write_random_snp --min_populations 1 -r 0.2

### populations output 
There are a ton of output files but the ones you really want to look at are the three *_summary.tsv files:

	ls -lh *_summary.tsv

Which lists:

	-rw-r--r--  1 kristenlehman  staff   716B Oct  4 11:09 CatalinaSIFI.p.fst_summary.tsv
	-rw-r--r--  1 kristenlehman  staff   1.5K Oct  4 11:09 CatalinaSIFI.p.phistats_summary.tsv
	-rw-r--r--  1 kristenlehman  staff   5.0K Oct  4 11:09 CatalinaSIFI.p.sumstats_summary.tsv

https://catchenlab.life.illinois.edu/stacks/manual/#pfiles gives tables of names for the stats columns headers to all output files.

# Run populations: between islands

### pop_file for between island population stats
`SIFI123a_stacks_pop_file.csv`

		Sibara_filifolia_AJC_9654cat_SCl_1	SCl
		Sibara_filifolia_AJC_9656cat_SCl_1	SCl
		Sibara_filifolia_AJC_Z1_2_9655	SCl
		Sibara_filifolia_AJC_Z1_4_9657	SCl
		Sibara_filifolia_AJC_Z1_5_9653	SCl
		Sibara_filifolia_AJK_9664cat_SCl_3	SCl
		Sibara_filifolia_AJK_Z3_2_9662	SCl
		Sibara_filifolia_AJK_Z3_5_9663	SCl
		Sibara_filifolia_HES_SIF123_Bunkers2_9611	SCa
		Sibara_filifolia_HES_SIFI111_9599	SCa
Between isl run: 10:30 start

	populations -V '/PATH/TO/.vcf' -O '/PATH/TO/_stacks_outfiles -M '/PATH/TO/_stacks_pop_file.csv --fstats --write_random_snp --min_populations 1 -r 0.2

### printed output

		Found 161240 SNP records in file '/PATH/TO/.vcf'. (Skipped 0 already filtered-out SNPs and 0 non-SNP records ; more with --verbose.)
		
		Removed 1633 loci that did not pass sample/population constraints from 161240 loci.
		Kept 159607 loci, composed of 159607 sites; 209 of those sites were filtered, 139833 variant sites remained.
		    157454 genomic sites, of which 80 were covered by multiple loci (0.1%).
		Mean genotyped sites per locus: 0.98bp (stderr 0.00).
		
		Population summary statistics (more detail in populations.sumstats_summary.tsv):
		  SCl: 52.207 samples per locus; pi: 0.042669; all/variant/polymorphic sites: 102742/88103/21639; private alleles: 6782
		  SCa: 67.815 samples per locus; pi: 0.084139; all/variant/polymorphic sites: 124550/121672/100273; private alleles: 50989
		  C_h: 1.3647 samples per locus; pi: 0.083807; all/variant/polymorphic sites: 20464/19169/1901; private alleles: 3178
		
		Population pair divergence statistics (more in populations.fst_summary.tsv and populations.phistats_summary.tsv):
		  SCl-SCa: mean Fst: 0.25733; mean Phi_st: 0.29074; mean Fst': 0.25649
		  SCl-C_h: mean Fst: 0.67513; mean Phi_st: 0.77442; mean Fst': 0.66448
		  SCa-C_h: mean Fst: 0.23712; mean Phi_st: 0.18066; mean Fst': 0.25555
		
		Populations is done.
# Run populations: within Clemente sites

### pop_codes
`ClemSIFI123a_stacks_pop_codes.csv`

		Sibara_filifolia_AJC_9654cat_SCl_1	Zone_1
		Sibara_filifolia_AJC_9656cat_SCl_1	Zone_1
		Sibara_filifolia_AJC_Z1_2_9655_SCl_1	Zone_1
		Sibara_filifolia_KHL4798_Z2_9659_SCl_2	Zone_2
		Sibara_filifolia_KHL4802_Z2_10269_SCl_2	Zone_2
		Sibara_filifolia_KHLM36_9645cat_SCl_4	Zone_4
  
### run populations
	populations -V /Users/kristenlehman/Desktop/Projects/SIFI/ipyrad_data_processing/ClementeSIFI_outfiles/ClementeSIFI.vcf -O /Users/kristenlehman/Desktop/Projects/SIFI/stacks/ClemSIFI123a/Clem_SIFI123a_stacks_outfiles -M /Users/kristenlehman/Desktop/Projects/SIFI/stacks/ClemSIFI123a/ClemSIFI123a_stacks_pop_codes.csv --fstats --write_random_snp --min_populations 2 -r 0.3

### printed output

If its running correctly, it'll stream `RAD_###` for a minute and finally print this:

		.
		.
		.
	 	RAD_35830
		RAD_35840
		RAD_35851
		RAD_35864
		
		Found 25040 SNP records in file '/Users/kristenlehman/Desktop/Projects/SIFI/ipyrad_data_processing/ClementeSIFI_outfiles/ClementeSIFI.vcf'. (Skipped 0 already filtered-out SNPs and 0 non-SNP records ; more with --verbose.)
		
		Removed 9809 loci that did not pass sample/population constraints from 25040 loci.
		Kept 15231 loci, composed of 15231 sites; 20 of those sites were filtered, 12939 variant sites remained.
		    15186 genomic sites, of which 45 were covered by multiple loci (0.3%).
		Mean genotyped sites per locus: 1.00bp (stderr 0.00).
		
		Population summary statistics (more detail in populations.sumstats_summary.tsv):
		  Zone_1: 8.7401 samples per locus; pi: 0.11212; all/variant/polymorphic sites: 13371/11574/4711; private alleles: 1104
		  Zone_3: 12.289 samples per locus; pi: 0.15049; all/variant/polymorphic sites: 11309/10152/4218; private alleles: 988
		  Zone_2: 8.9624 samples per locus; pi: 0.18963; all/variant/polymorphic sites: 13972/12024/6172; private alleles: 1220
		  Zone_4: 15.034 samples per locus; pi: 0.049445; all/variant/polymorphic sites: 14349/12300/5184; private alleles: 883
		  Zone_5: 12.886 samples per locus; pi: 0.12095; all/variant/polymorphic sites: 5553/5325/1967; private alleles: 345
		  Zone_6: 3.9337 samples per locus; pi: 0.011172; all/variant/polymorphic sites: 7558/7335/251; private alleles: 1856
		
		Population pair divergence statistics (more in populations.fst_summary.tsv and populations.phistats_summary.tsv):
		  Zone_1-Zone_3: mean Fst: 0.23039; mean Phi_st: 0.28628; mean Fst': 0.23261
		  Zone_1-Zone_2: mean Fst: 0.16326; mean Phi_st: 0.20233; mean Fst': 0.15773
		  Zone_1-Zone_4: mean Fst: 0.31446; mean Phi_st: 0.35568; mean Fst': 0.32278
		  Zone_1-Zone_5: mean Fst: 0.17011; mean Phi_st: 0.22783; mean Fst': 0.20125
		  Zone_1-Zone_6: mean Fst: 0.5751; mean Phi_st: 0.59792; mean Fst': 0.60126
		  Zone_3-Zone_2: mean Fst: 0.1266; mean Phi_st: 0.16434; mean Fst': 0.10019
		  Zone_3-Zone_4: mean Fst: 0.15946; mean Phi_st: 0.20488; mean Fst': 0.14673
		  Zone_3-Zone_5: mean Fst: 0.23463; mean Phi_st: 0.29977; mean Fst': 0.23311
		  Zone_3-Zone_6: mean Fst: 0.51703; mean Phi_st: 0.56486; mean Fst': 0.55424
		  Zone_2-Zone_4: mean Fst: 0.18058; mean Phi_st: 0.23869; mean Fst': 0.1594
		  Zone_2-Zone_5: mean Fst: 0.18059; mean Phi_st: 0.23975; mean Fst': 0.18367
		  Zone_2-Zone_6: mean Fst: 0.50558; mean Phi_st: 0.5425; mean Fst': 0.52581
		  Zone_4-Zone_5: mean Fst: 0.2988; mean Phi_st: 0.36477; mean Fst': 0.31161
		  Zone_4-Zone_6: mean Fst: 0.56977; mean Phi_st: 0.58846; mean Fst': 0.60437
		  Zone_5-Zone_6: mean Fst: 0.54122; mean Phi_st: 0.57956; mean Fst': 0.57413
	
			Populations is done.


